### PR TITLE
UI Feature: Global search across tx, addresses, amounts, xpubs

### PR DESCRIPTION
**Pull request by relativisticelectron**
See: https://github.com/cryptoadvance/specter-desktop/issues/1593

For me most important is the feature to quickly search all wallets for an address.

Peek 2022-07-11 16-33

TODO:
- [x] https://github.com/cryptoadvance/specter-desktop/issues/1794 has to be fixed first, so searches in addresses work
- [ ] send cleaner json via request 
- [ ] Prevent searching in tx fields, that are not visible to the user in tx-table
- [ ] Make the search entry persistent when switching sites,  or maybe have search results in a clickable dropdown list?

If you have **feedback**, e.g., how to make the UI nicer, let me know.  
- Currently I do not know how to make the search_term persist when loading a different page

### CODE REVIEW

**Comment by relativisticelectron** [comment]
Moved the search field to the sidebar
image

**Comment by k9ert** [comment]
WOW, this looks awesome! Will have a look soon!

File: src/cryptoadvance/specter/global_search.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
+import os
+import logging
+
+from flask import url_for
+
+logger = logging.getLogger(__name__)
+
+
+class HtmlElement:
```
Comment (k9ert) [inline src/cryptoadvance/specter/global_search.py:9]: This sounds cool but i don't have an idea from that description. Maybe it's a good candidate for some tests? It looks like it's quite easy to write some tests for it and that would give a much better picture how this works.

File: src/cryptoadvance/specter/helpers.py [hunk 2]
```diff
@@ -307,3 +308,17 @@ def get_address_from_dict(data_dict):
     if addr and addr != "Fee":
         return addr
     raise RuntimeError(f"Missing address info in object {data_dict}")
+
+
+def robust_json_dumps(obj):
```
Comment (k9ert) [inline src/cryptoadvance/specter/helpers.py:313]: Good point. json parsing is sometimes quite annoying. But can you move that to somewhere in util? `util/json_rpc.py` would do it or a new `util/misc.py`. At some point, all that stuff in the util folder was initially in `helpers.py` and hopefully at some point the `helpers.py` vanish.

Maybe you can add a deprecation warning like 
```
Additions to helpers.py are deprecated. Please place your helpers in utils/somewhere
```

Comment (relativisticelectron) [reply src/cryptoadvance/specter/helpers.py]: Done

File: src/cryptoadvance/specter/global_search.py (hunk 1 above)
Comment (k9ert) [reply src/cryptoadvance/specter/global_search.py:9]: Preserving the search-term should be easy via storing it in the session?! I try to get out of the way of UI judgement. Don't have a good track record there. @moneymanolis  @b30wulffz what do you think?

File: src/cryptoadvance/specter/global_search.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    def result(self):
+        if self._result:
+            return self._result
+        else:
+            return sum([child.result for child in self.children])
+
+    @result.setter
+    def result(self, result):
+        if self.children:
+            raise "Setting results is only allowed for end nodes"
```
Comment (k9ert) [inline src/cryptoadvance/specter/global_search.py:41]: Raising a string? Shouldn't you raise an Exception or a descendant from an exception?

File: src/cryptoadvance/specter/global_search.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    "Given an html_root it will call the child.function for all childs that do not have any children"
+    end_nodes = html_root.calculate_end_nodes()
+    for end_node in end_nodes:
+        end_node.result = end_node.function(search_term)
+    return html_root
+
+
+def do_global_search(search_term, specter):
+    "Builds the HTML Tree if ncessary (only do it once) and then calls the functions in it to search for the search_term"
+    global HTML_ROOT
```
Comment (k9ert) [inline src/cryptoadvance/specter/global_search.py:246]: I'm not sure that's a good idea?! Is that possible with multi-user? How expensive is the build-up of the tree? Maybe we can do it every time?

File: src/cryptoadvance/specter/global_search.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    It also encodes which functions will be used for searching.
+
+    Returns:
+        HtmlElement: This is the html_root, which has all children linked inside
+    """
+    html_root = HtmlElement(None)
+    wallets = HtmlElement(html_root, id="toggle_wallets_list")
+    devices = HtmlElement(html_root, id="toggle_devices_list")
+
+    def search_in_structure(search_term, l):
```
Comment (k9ert) [inline src/cryptoadvance/specter/global_search.py:89]: Puh, this function in function-scope is imho quite horrible. Especially evil is the example below where you're not even returning a value but just changing variables which are in the scope. To me, this is a very confusing programming-model.

Shouldn't it be possible to include these methods in the HtmlElement-class?

File: src/cryptoadvance/specter/global_search.py (hunk 4 above)
Comment (relativisticelectron) [reply src/cryptoadvance/specter/global_search.py:246]: Yes, you are right, I didn't think about multi-user yet.  Please do not put in too much effort now to review the code. Before I put more work in the code (and have to change a lot), I am seeking feedback, if this  feature is wanted, and how the UI should look like.

File: src/cryptoadvance/specter/global_search.py (hunk 5 above)
Comment (relativisticelectron) [reply src/cryptoadvance/specter/global_search.py:89]: I agree, this needs a restructuring.

**Comment by moneymanolis** [comment]
@relativisticelectron I'd need to first understand the goals associated with this PR before we can talk UI options etc. What do you want to search for? When is the global search used / sensible and when the search bar of the tx table?

So far, I could only gather: "feature to quickly search all wallets for an address." - which is fine if you want to double check whether an address belongs to one of your wallets and you don't remember which one. 

My rough first intuition is, that we talk about a two step approach here. (1) The global search tells you in which wallet(s) to look for, (2) and then, say, the search in the tx table takes over

**Comment by relativisticelectron** [comment]
> @relativisticelectron I'd need to first understand the goals associated with this PR before we can talk UI options etc. What do you want to search for? When is the global search used / sensible and when the search bar of the tx table?
> 
> So far, I could only gather: "feature to quickly search all wallets for an address." - which is fine if you want to double check whether an address belongs to one of your wallets and you don't remember which one.
> 
> My rough first intuition is, that we talk about a two step approach here. (1) The global search tells you in which wallet(s) to look for, (2) and then, say, the search in the tx table takes over

The original inspiration is gmail. The search searches everything, is fast and changes the experience/behavior of "I have to remember in which folder the email is" to "just search for it".   **UX: Search (and see results immediately) -->  Click on result --> See result fully.  --> Go back to result list if it was the wrong one**
Similar fantastic search is in VS Code for files and settings. 

In Specter I had the experience: In which wallet was the address and was it used already?  I would like to mirror the gmail UX flow with as few steps as possible (I think the current PR doesn't do that, because there are too many clicks, but it was a first attempt).
And if we do it for addresses, there is no reason to not also search transactions ids, labels, amounts, xpubs, ...

**Comment by relativisticelectron** [comment]
Interesting.  Currently the "History" and "UTXO"   however doesn't show unused addresses belonging to a wallet. And adding it seems kind of wrong, because then it is not the   "History" any more.

I think most users are used to global searches not attached to a certain table but in the upper left/right, e.g.
image

**Comment by relativisticelectron** [comment]
I have to redesign this search. Currently getting to the result needs way too many clicks.... Closing this PR for now.

If anyone knows a good webcomponent for a rich-text dropdown, then please write it here.

**Pull request closed without merging**